Bug: watch mode does not work since 2.8.0-dev.20180314

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨 -->
<!--
Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker.

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the CONTRIBUTING guidelines: https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ
-->

<!-- If you have a QUESTION:
   THIS IS NOT A FORUM FOR QUESTIONS.
   Ask questions at http://stackoverflow.com/questions/tagged/typescript
    or https://gitter.im/Microsoft/TypeScript
-->

<!-- If you have a SUGGESTION:
  Most suggestion reports are duplicates, please search extra hard before logging a new suggestion.
  See https://github.com/Microsoft/TypeScript-wiki/blob/master/Writing-Good-Design-Proposals.md
-->

<!-- If you have a BUG:
  Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  2.8.0-dev.20180316

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** 

**Code**

tsc -w

**Expected behavior:**

Starting compilation in watch mode...
Compilation complete. Watching for file changes.

**Actual behavior:**
Starting compilation in watch mode...

and no files generated

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior --> 

**Related Issues:**

@taoqf What are the exact steps you are following. 
Here is what I did with your sample project
```
c:\github\sample>node c:\TypeScript\built\local\tsc.js -w --preserveWatchOutput
[09:24:12] Starting compilation in watch mode...


src/sample.ts:1:1 - error TS6133: 'fun' is declared but its value is never read.

1 import { fun } from 'not-exist';
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


src/sample.ts:1:21 - error TS2307: Cannot find module 'not-exist'.

1 import { fun } from 'not-exist';
                      ~~~~~~~~~~~


src/sample.ts:3:7 - error TS6133: 'no' is declared but its value is never read.

3 const no = 'fd-w000001';
        ~~


[09:24:13] Compilation complete. Watching for file changes.
```

Ran ````npm install``` after this from another command prompt. There is no additional output since there is no change in program with those installs..

Changed ```no``` to ```no2``` and I see new errors being presented as:
```
[09:26:02] File change detected. Starting incremental compilation...


src/sample.ts:1:1 - error TS6133: 'fun' is declared but its value is never read.

1 import { fun } from 'not-exist';
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


src/sample.ts:1:21 - error TS2307: Cannot find module 'not-exist'.

1 import { fun } from 'not-exist';
                      ~~~~~~~~~~~


src/sample.ts:3:7 - error TS6133: 'no2' is declared but its value is never read.

3 const no2 = 'fd-w000001';
        ~~~


[09:26:02] Compilation complete. Watching for file changes.
```

New outputs are also generated correctly and don't see any impact on cpu cycles.

Is there something I am missing.

Update I added changed code to import from webpack-stream and it looks like error isn't being reported correctly and looking into it.. though it seems like it isn't just watch mode issue since it happens without --watch
```
c:\github\sample>node c:\TypeScript\built\local\tsc.js
======== Resolving module 'webpack-stream' from 'c:/github/sample/src/sample.ts'. ========
Explicitly specified module resolution kind: 'NodeJs'.
Loading module 'webpack-stream' from 'node_modules' folder, target file type 'TypeScript'.
Directory 'c:/github/sample/src/node_modules' does not exist, skipping all lookups in it.
Directory 'c:/github/sample/node_modules' does not exist, skipping all lookups in it.
Directory 'c:/github/node_modules' does not exist, skipping all lookups in it.
Directory 'c:/node_modules' does not exist, skipping all lookups in it.
Loading module 'webpack-stream' from 'node_modules' folder, target file type 'JavaScript'.
Directory 'c:/github/sample/src/node_modules' does not exist, skipping all lookups in it.
Directory 'c:/github/sample/node_modules' does not exist, skipping all lookups in it.
Directory 'c:/github/node_modules' does not exist, skipping all lookups in it.
Directory 'c:/node_modules' does not exist, skipping all lookups in it.
======== Module name 'webpack-stream' was not resolved. ========
======== Resolving module 'not-exist' from 'c:/github/sample/src/sample.ts'. ========
Explicitly specified module resolution kind: 'NodeJs'.
Loading module 'not-exist' from 'node_modules' folder, target file type 'TypeScript'.
Directory 'c:/github/sample/src/node_modules' does not exist, skipping all lookups in it.
Directory 'c:/github/sample/node_modules' does not exist, skipping all lookups in it.
Directory 'c:/github/node_modules' does not exist, skipping all lookups in it.
Directory 'c:/node_modules' does not exist, skipping all lookups in it.
Loading module 'not-exist' from 'node_modules' folder, target file type 'JavaScript'.
Directory 'c:/github/sample/src/node_modules' does not exist, skipping all lookups in it.
Directory 'c:/github/sample/node_modules' does not exist, skipping all lookups in it.
Directory 'c:/github/node_modules' does not exist, skipping all lookups in it.
Directory 'c:/node_modules' does not exist, skipping all lookups in it.
======== Module name 'not-exist' was not resolved. ========
src/sample.ts:3:1 - error TS6133: 'fun' is declared but its value is never read.

3 import { fun } from 'not-exist';
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


src/sample.ts:3:21 - error TS2307: Cannot find module 'not-exist'.

3 import { fun } from 'not-exist';
                      ~~~~~~~~~~~


src/sample.ts:5:7 - error TS6133: 'no2' is declared but its value is never read.

5 const no2 = 'fd-w000001';
        ~~~
```

Followed your steps and yet dont see issue with cpu or reporting error or file generation?
```
c:\github\sample>npm i -g cnpm
npm WARN deprecated socks@1.1.10: If using 2.x branch, please upgrade to at least 2.1.6 to avoid a serious bug with socket data flow and an import issue introduced in 2.1.0
C:\Users\shkamat\AppData\Roaming\npm\cnpm -> C:\Users\shkamat\AppData\Roaming\npm\node_modules\cnpm\bin\cnpm
+ cnpm@5.2.0
added 778 packages in 38.879s

c:\github\sample>cnpm i
/ [1/2] Installing static-extend@^0.1.1platform unsupported webpack-stream@4.0.2 › webpack@3.11.0 › watchpack@1.5.0 › chokidar@2.0.3 › fsevents@^1.1.2 Package require os(darwin) not compatible with your platform(win32)
[fsevents@^1.1.2] optional install error: Package require os(darwin) not compatible with your platform(win32)
√ Installed 2 packages
√ Linked 294 latest versions
√ Run 1 scripts
Recently updated (since 2018-03-16): 7 packages (detail see file c:\github\sample\node_modules\.recently_updates.txt)
√ All packages installed (327 packages installed from npm registry, used 27s, speed 572.18kB/s, json 296(4.01MB), tarball 11.21MB)

c:\github\sample>node_modules\.bin\tsc --w --preserveWatchOutput
[10:38:07] Starting compilation in watch mode...


======== Resolving module 'not-exist' from 'c:/github/sample/src/sample.ts'. ========
Explicitly specified module resolution kind: 'NodeJs'.
Loading module 'not-exist' from 'node_modules' folder, target file type 'TypeScript'.
Directory 'c:/github/sample/src/node_modules' does not exist, skipping all lookups in it.
File 'c:/github/sample/node_modules/not-exist.ts' does not exist.
File 'c:/github/sample/node_modules/not-exist.tsx' does not exist.
File 'c:/github/sample/node_modules/not-exist.d.ts' does not exist.
Directory 'c:/github/sample/node_modules/@types' does not exist, skipping all lookups in it.
Directory 'c:/github/node_modules' does not exist, skipping all lookups in it.
Directory 'c:/node_modules' does not exist, skipping all lookups in it.
Loading module 'not-exist' from 'node_modules' folder, target file type 'JavaScript'.
Directory 'c:/github/sample/src/node_modules' does not exist, skipping all lookups in it.
File 'c:/github/sample/node_modules/not-exist.js' does not exist.
File 'c:/github/sample/node_modules/not-exist.jsx' does not exist.
Directory 'c:/github/node_modules' does not exist, skipping all lookups in it.
Directory 'c:/node_modules' does not exist, skipping all lookups in it.
======== Module name 'not-exist' was not resolved. ========
src/sample.ts:1:1 - error TS6133: 'fun' is declared but its value is never read.

1 import { fun } from 'not-exist';
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


src/sample.ts:1:21 - error TS2307: Cannot find module 'not-exist'.

1 import { fun } from 'not-exist';
                      ~~~~~~~~~~~


src/sample.ts:3:7 - error TS6133: 'no' is declared but its value is never read.

3 const no = 'fd-w000001';
        ~~


[10:38:08] Compilation complete. Watching for file changes.


[10:39:26] File change detected. Starting incremental compilation...


src/sample.ts:1:1 - error TS6133: 'fun2' is declared but its value is never read.

1 import { fun2 } from 'not-exist';
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


src/sample.ts:1:22 - error TS2307: Cannot find module 'not-exist'.

1 import { fun2 } from 'not-exist';
                       ~~~~~~~~~~~


src/sample.ts:3:7 - error TS6133: 'no' is declared but its value is never read.

3 const no = 'fd-w000001';
        ~~


[10:39:26] Compilation complete. Watching for file changes.
```

--- a/src/compiler/sys.ts
+++ b/src/compiler/sys.ts
@@ -337,6 +337,7 @@ namespace ts {
         getAccessileSortedChildDirectories(path: string): ReadonlyArray<string>;
         directoryExists(dir: string): boolean;
         filePathComparer: Comparer<string>;
+        realpath(s: string): string;
     }
 
     /**
@@ -392,9 +393,12 @@ namespace ts {
         function watchChildDirectories(parentDir: string, existingChildWatches: ChildWatches, callback: DirectoryWatcherCallback): ChildWatches {
             let newChildWatches: DirectoryWatcher[] | undefined;
             enumerateInsertsAndDeletes<string, DirectoryWatcher>(
-                host.directoryExists(parentDir) ? host.getAccessileSortedChildDirectories(parentDir) : emptyArray,
+                host.directoryExists(parentDir) ? mapDefined(host.getAccessileSortedChildDirectories(parentDir), child => {
+                    const childFullName = getNormalizedAbsolutePath(child, parentDir);
+                    return host.filePathComparer(childFullName, host.realpath(childFullName)) === Comparison.EqualTo ? childFullName : undefined;
+                }) : emptyArray,
                 existingChildWatches,
-                (child, childWatcher) => host.filePathComparer(getNormalizedAbsolutePath(child, parentDir), childWatcher.dirName),
+                (child, childWatcher) => host.filePathComparer(child, childWatcher.dirName),
                 createAndAddChildDirectoryWatcher,
                 closeFileWatcher,
                 addChildDirectoryWatcher
@@ -406,7 +410,7 @@ namespace ts {
              * Create new childDirectoryWatcher and add it to the new ChildDirectoryWatcher list
              */
             function createAndAddChildDirectoryWatcher(childName: string) {
-                const result = createDirectoryWatcher(getNormalizedAbsolutePath(childName, parentDir), callback);
+                const result = createDirectoryWatcher(childName, callback);
                 addChildDirectoryWatcher(result);
             }
 
@@ -601,14 +605,7 @@ namespace ts {
                 exit(exitCode?: number): void {
                     process.exit(exitCode);
                 },
-                realpath(path: string): string {
-                    try {
-                        return _fs.realpathSync(path);
-                    }
-                    catch {
-                        return path;
-                    }
-                },
+                realpath,
                 debugMode: some(<string[]>process.execArgv, arg => /^--(inspect|debug)(-brk)?(=\d+)?$/i.test(arg)),
                 tryEnableSourceMapsForHost() {
                     try {
@@ -700,7 +697,8 @@ namespace ts {
                     filePathComparer: useCaseSensitiveFileNames ? compareStringsCaseSensitive : compareStringsCaseInsensitive,
                     directoryExists,
                     getAccessileSortedChildDirectories: path => getAccessibleFileSystemEntries(path).directories,
-                    watchDirectory
+                    watchDirectory,
+                    realpath
                 });
 
                 return (directoryName, callback, recursive) => {
@@ -1041,6 +1039,15 @@ namespace ts {
 
             function getDirectories(path: string): string[] {
                 return filter<string>(_fs.readdirSync(path), dir => fileSystemEntryExists(combinePaths(path, dir), FileSystemEntryKind.Directory));
+            }
+
+            function realpath(path: string): string {
+                try {
+                    return _fs.realpathSync(path);
+                }
+                catch {
+                    return path;
+                }
             }
 
             function getModifiedTime(path: string) {

--- a/src/compiler/sys.ts
+++ b/src/compiler/sys.ts
@@ -334,7 +334,7 @@ namespace ts {
     /*@internal*/
     export interface RecursiveDirectoryWatcherHost {
         watchDirectory: HostWatchDirectory;
-        getAccessileSortedChildDirectories(path: string): ReadonlyArray<string>;
+        getAccessibleSortedChildDirectories(path: string): ReadonlyArray<string>;
         directoryExists(dir: string): boolean;
         filePathComparer: Comparer<string>;
         realpath(s: string): string;
@@ -393,8 +393,10 @@ namespace ts {
         function watchChildDirectories(parentDir: string, existingChildWatches: ChildWatches, callback: DirectoryWatcherCallback): ChildWatches {
             let newChildWatches: DirectoryWatcher[] | undefined;
             enumerateInsertsAndDeletes<string, DirectoryWatcher>(
-                host.directoryExists(parentDir) ? mapDefined(host.getAccessileSortedChildDirectories(parentDir), child => {
+                host.directoryExists(parentDir) ? mapDefined(host.getAccessibleSortedChildDirectories(parentDir), child => {
                     const childFullName = getNormalizedAbsolutePath(child, parentDir);
+                    // Filter our the symbolic link directories since those arent included in recursive watch
+                    // which is same behaviour when recursive: true is passed to fs.watch
                     return host.filePathComparer(childFullName, host.realpath(childFullName)) === Comparison.EqualTo ? childFullName : undefined;
                 }) : emptyArray,
                 existingChildWatches,
@@ -696,7 +698,7 @@ namespace ts {
                 const watchDirectoryRecursively = createRecursiveDirectoryWatcher({
                     filePathComparer: useCaseSensitiveFileNames ? compareStringsCaseSensitive : compareStringsCaseInsensitive,
                     directoryExists,
-                    getAccessileSortedChildDirectories: path => getAccessibleFileSystemEntries(path).directories,
+                    getAccessibleSortedChildDirectories: path => getAccessibleFileSystemEntries(path).directories,
                     watchDirectory,
                     realpath
                 });

--- a/src/harness/unittests/tscWatchMode.ts
+++ b/src/harness/unittests/tscWatchMode.ts
@@ -2358,6 +2358,50 @@ declare module "fs" {
             it("uses non recursive dynamic polling when renaming file in subfolder", () => {
                 verifyRenamingFileInSubFolder(TestFSWithWatch.Tsc_WatchDirectory.DynamicPolling);
             });
+
+            it("when there are symlinks to folders in recursive folders", () => {
+                const cwd = "/home/user/projects/myproject";
+                const file1: FileOrFolder = {
+                    path: `${cwd}/src/file.ts`,
+                    content: `import * as a from "a"`
+                };
+                const tsconfig: FileOrFolder = {
+                    path: `${cwd}/tsconfig.json`,
+                    content: `{ "compilerOptions": { "extendedDiagnostics": true, "traceResolution": true }}`
+                };
+                const realA: FileOrFolder = {
+                    path: `${cwd}/node_modules/reala/index.d.ts`,
+                    content: `export {}`
+                };
+                const realB: FileOrFolder = {
+                    path: `${cwd}/node_modules/realb/index.d.ts`,
+                    content: `export {}`
+                };
+                const symLinkA: FileOrFolder = {
+                    path: `${cwd}/node_modules/a`,
+                    symLink: `${cwd}/node_modules/reala`
+                };
+                const symLinkB: FileOrFolder = {
+                    path: `${cwd}/node_modules/b`,
+                    symLink: `${cwd}/node_modules/realb`
+                };
+                const symLinkBInA: FileOrFolder = {
+                    path: `${cwd}/node_modules/reala/node_modules/b`,
+                    symLink: `${cwd}/node_modules/b`
+                };
+                const symLinkAInB: FileOrFolder = {
+                    path: `${cwd}/node_modules/realb/node_modules/a`,
+                    symLink: `${cwd}/node_modules/a`
+                };
+                const files = [file1, tsconfig, realA, realB, symLinkA, symLinkB, symLinkBInA, symLinkAInB];
+                const environmentVariables = createMap<string>();
+                environmentVariables.set("TSC_WATCHDIRECTORY", TestFSWithWatch.Tsc_WatchDirectory.NonRecursiveWatchDirectory);
+                const host = createWatchedSystem(files, { environmentVariables, currentDirectory: cwd });
+                createWatchOfConfigFile("tsconfig.json", host);
+                checkWatchedDirectories(host, emptyArray, /*recursive*/ true);
+                checkWatchedDirectories(host, [cwd, `${cwd}/node_modules`, `${cwd}/node_modules/@types`, `${cwd}/node_modules/reala`, `${cwd}/node_modules/realb`,
+                    `${cwd}/node_modules/reala/node_modules`, `${cwd}/node_modules/realb/node_modules`, `${cwd}/src`], /*recursive*/ false);
+            });
         });
     });
 }

--- a/src/harness/virtualFileSystemWithWatch.ts
+++ b/src/harness/virtualFileSystemWithWatch.ts
@@ -314,7 +314,8 @@ interface Array<T> {}`
                     directoryExists: path => this.directoryExists(path),
                     getAccessileSortedChildDirectories: path => this.getDirectories(path),
                     filePathComparer: this.useCaseSensitiveFileNames ? compareStringsCaseSensitive : compareStringsCaseInsensitive,
-                    watchDirectory
+                    watchDirectory,
+                    realpath: s => this.realpath(s)
                 });
             }
             else if (tscWatchDirectory === Tsc_WatchDirectory.NonRecursiveWatchDirectory) {
@@ -323,7 +324,8 @@ interface Array<T> {}`
                     directoryExists: path => this.directoryExists(path),
                     getAccessileSortedChildDirectories: path => this.getDirectories(path),
                     filePathComparer: this.useCaseSensitiveFileNames ? compareStringsCaseSensitive : compareStringsCaseInsensitive,
-                    watchDirectory
+                    watchDirectory,
+                    realpath: s => this.realpath(s)
                 });
             }
             else if (tscWatchDirectory === Tsc_WatchDirectory.DynamicPolling) {
@@ -333,7 +335,8 @@ interface Array<T> {}`
                     directoryExists: path => this.directoryExists(path),
                     getAccessileSortedChildDirectories: path => this.getDirectories(path),
                     filePathComparer: this.useCaseSensitiveFileNames ? compareStringsCaseSensitive : compareStringsCaseInsensitive,
-                    watchDirectory
+                    watchDirectory,
+                    realpath: s => this.realpath(s)
                 });
             }
         }
@@ -649,7 +652,7 @@ interface Array<T> {}`
 
             const realpath = this.realpath(path);
             if (path !== realpath) {
-                return this.getRealFsEntry(isFsEntry, realpath as Path);
+                return this.getRealFsEntry(isFsEntry, this.toPath(realpath));
             }
 
             return undefined;

--- a/src/harness/virtualFileSystemWithWatch.ts
+++ b/src/harness/virtualFileSystemWithWatch.ts
@@ -312,7 +312,7 @@ interface Array<T> {}`
                 const watchDirectory: HostWatchDirectory = (directory, cb) => this.watchFile(directory, () => cb(directory), PollingInterval.Medium);
                 this.customRecursiveWatchDirectory = createRecursiveDirectoryWatcher({
                     directoryExists: path => this.directoryExists(path),
-                    getAccessileSortedChildDirectories: path => this.getDirectories(path),
+                    getAccessibleSortedChildDirectories: path => this.getDirectories(path),
                     filePathComparer: this.useCaseSensitiveFileNames ? compareStringsCaseSensitive : compareStringsCaseInsensitive,
                     watchDirectory,
                     realpath: s => this.realpath(s)
@@ -322,7 +322,7 @@ interface Array<T> {}`
                 const watchDirectory: HostWatchDirectory = (directory, cb) => this.watchDirectory(directory, fileName => cb(fileName), /*recursive*/ false);
                 this.customRecursiveWatchDirectory = createRecursiveDirectoryWatcher({
                     directoryExists: path => this.directoryExists(path),
-                    getAccessileSortedChildDirectories: path => this.getDirectories(path),
+                    getAccessibleSortedChildDirectories: path => this.getDirectories(path),
                     filePathComparer: this.useCaseSensitiveFileNames ? compareStringsCaseSensitive : compareStringsCaseInsensitive,
                     watchDirectory,
                     realpath: s => this.realpath(s)
@@ -333,7 +333,7 @@ interface Array<T> {}`
                 const watchDirectory: HostWatchDirectory = (directory, cb) => watchFile(directory, () => cb(directory), PollingInterval.Medium);
                 this.customRecursiveWatchDirectory = createRecursiveDirectoryWatcher({
                     directoryExists: path => this.directoryExists(path),
-                    getAccessileSortedChildDirectories: path => this.getDirectories(path),
+                    getAccessibleSortedChildDirectories: path => this.getDirectories(path),
                     filePathComparer: this.useCaseSensitiveFileNames ? compareStringsCaseSensitive : compareStringsCaseInsensitive,
                     watchDirectory,
                     realpath: s => this.realpath(s)

